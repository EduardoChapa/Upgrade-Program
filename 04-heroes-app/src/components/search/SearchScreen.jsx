import React, { useMemo } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import HeroCard from '../hero/HeroCard';
import getHeroesByName from '../selectors/getHeroesByName';
import queryString from 'query-string';

function SearchScreen() {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);
    
    const [ formValues, handleInputChange ] = useForm({ search: q });
    
    const { search } = formValues;
    
    const heroes = useMemo(() => getHeroesByName(q), [q]);
    
    const handleSearch = (e) => {
        e.preventDefault();
        navigate(`?q=${ search }`);
    }
    

    return <div>
        <h1>Search Screen!</h1>
        <hr/>

        <div className='row'>
            <div className='col-5'>
                <h4>Search</h4>
                <hr/>

                <form onSubmit={ handleSearch }>
                    <input type="text" placeholder='Search a hero...' className='form-control' name='search' autoComplete='off' value={ search } onChange={ handleInputChange }/>
                    <button className='btn btn-outline-primary mt-2' type='submit'>Search</button>
                </form>
            </div>

            <div className='col-7'>
                <h4>Results</h4>
                <hr/>

                {
                    (q === '') && 
                    <div className='alert alert-info'>
                        Search for a hero
                    </div>
                }

                {
                    heroes.map(hero => (
                        <HeroCard key={ hero.id } { ...hero }/>
                    ))
                }
            </div>
        </div>
    </div>
}

export default SearchScreen;
