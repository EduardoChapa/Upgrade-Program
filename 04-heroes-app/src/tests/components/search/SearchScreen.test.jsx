import SearchScreen from "../../../components/search/SearchScreen";
import { MemoryRouter } from "react-router-dom";
import { mount } from 'enzyme';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));

describe('Testing SearchScreen component', () => {
    test('should display correctly with default values', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>    
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search for a hero...');
    });

    test('should show batman and the input should have query value', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>    
        );
        expect(wrapper.find('input').prop('value')).toBe('batman');
    });
    
    test('should display an error when no heroes were found', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=ohno']}>
                <SearchScreen />
            </MemoryRouter>    
        );
        expect(wrapper.find('input').prop('value')).toBe('ohno');
        expect(wrapper.find('.alert-danger').exists()).toBe(true);       
    });

    test('should call navigate to the correct url', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=ohno']}>
                <SearchScreen />
            </MemoryRouter>    
        );
        wrapper.find('input').simulate('change', {
            target: {
                name: 'search',
                value: 'batman'
            }
        });
        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {}
        });
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
    });
});