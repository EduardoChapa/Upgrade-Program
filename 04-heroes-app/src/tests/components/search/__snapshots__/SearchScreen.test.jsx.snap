// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SearchScreen component should display correctly with default values 1`] = `
<div>
  <h1>
    Search Screen!
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Search a hero..."
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-2"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search for a hero...
      </div>
    </div>
  </div>
</div>
`;
